File: langs/typescript/src/conformance/runner.ts
/**
 * Conformance runner: for each spec/conformance/scenarios/*.json,
 * load, solve, run both analyses, compare against expected outputs.
 *
 * Exit 0 on pass, non-zero on divergence.
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadScenario } from '../models/scenario-loader.js';
import { solve } from '../models/topsis/solve.js';
import { criticalDecisions } from '../models/topsis/critical-decisions.js';
import { criticalConstraints } from '../models/topsis/critical-constraints.js';
import type { CandidateM } from '../models/candidate.js';
import type { CriticalDecisionM } from '../models/critical-decision.js';
import type { CriticalConstraintM } from '../models/critical-constraint.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

function findRepoRoot(startDir: string): string {
  let dir = startDir;
  while (true) {
    if (fs.existsSync(path.join(dir, '.git'))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) throw new Error('Could not find repo root (.git) from ' + startDir);
    dir = parent;
  }
}

const REPO_ROOT = findRepoRoot(__dirname);
const SCENARIOS_DIR = path.join(REPO_ROOT, 'spec', 'conformance', 'scenarios');
const EXPECTED_DIR = path.join(REPO_ROOT, 'spec', 'conformance', 'expected');
const TOLERANCES_PATH = path.join(REPO_ROOT, 'spec', 'conformance', 'tolerances.json');

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const tolerances: any = JSON.parse(fs.readFileSync(TOLERANCES_PATH, 'utf-8'));
const SCALAR_TOL: number = tolerances.candidates.score.absolute as number; // 1e-9

// ---------------------------------------------------------------------------
// Comparison helpers
// ---------------------------------------------------------------------------
interface Divergence {
  field: string;
  expected: unknown;
  actual: unknown;
  diff?: number;
}

function closeEnough(a: number, b: number): boolean {
  return Math.abs(a - b) <= SCALAR_TOL;
}

function compareCandidates(
  actual: CandidateM[],
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  expected: any[],
  label: string,
): Divergence[] {
  const divs: Divergence[] = [];

  if (actual.length !== expected.length) {
    divs.push({
      field: `${label}.length`,
      expected: expected.length,
      actual: actual.length,
    });
    return divs;
  }

  for (let i = 0; i < actual.length; i++) {
    const a = actual[i];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const e = expected[i] as any;
    const prefix = `${label}[${i}]`;

    // alternativeIds
    const eIds = e.alternativeIds as string[];
    if (a.alternativeIds.length !== eIds.length) {
      divs.push({
        field: `${prefix}.alternativeIds.length`,
        expected: eIds.length,
        actual: a.alternativeIds.length,
      });
    } else {
      for (let j = 0; j < eIds.length; j++) {
        if (a.alternativeIds[j] !== eIds[j]) {
          divs.push({
            field: `${prefix}.alternativeIds[${j}]`,
            expected: eIds[j],
            actual: a.alternativeIds[j],
          });
        }
      }
    }

    // triangularValue
    for (const k of ['lower', 'modal', 'upper'] as const) {
      const av = a.triangularValue[k];
      const ev = e.triangularValue[k] as number;
      if (!closeEnough(av, ev)) {
        divs.push({
          field: `${prefix}.triangularValue.${k}`,
          expected: ev,
          actual: av,
          diff: Math.abs(av - ev),
        });
      }
    }

    // normalizedValue
    for (const k of ['positive', 'average', 'negative'] as const) {
      const av = a.normalizedValue[k];
      const ev = e.normalizedValue[k] as number;
      if (!closeEnough(av, ev)) {
        divs.push({
          field: `${prefix}.normalizedValue.${k}`,
          expected: ev,
          actual: av,
          diff: Math.abs(av - ev),
        });
      }
    }

    // score
    if (!closeEnough(a.score, e.score as number)) {
      divs.push({
        field: `${prefix}.score`,
        expected: e.score,
        actual: a.score,
        diff: Math.abs(a.score - (e.score as number)),
      });
    }

    // rank (exact)
    if (a.rank !== (e.rank as number)) {
      divs.push({ field: `${prefix}.rank`, expected: e.rank, actual: a.rank });
    }
  }
  return divs;
}

function compareCriticalDecisions(
  actual: CriticalDecisionM[],
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  expected: any[],
  label: string,
): Divergence[] {
  const divs: Divergence[] = [];

  if (actual.length !== expected.length) {
    divs.push({ field: `${label}.length`, expected: expected.length, actual: actual.length });
    return divs;
  }

  for (let i = 0; i < actual.length; i++) {
    const a = actual[i];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const e = expected[i] as any;
    const prefix = `${label}[${i}]`;

    if (a.decisionId !== (e.decisionId as string)) {
      divs.push({ field: `${prefix}.decisionId`, expected: e.decisionId, actual: a.decisionId });
    }
    for (const k of ['lower', 'modal', 'upper'] as const) {
      const av = a.triangularValue[k];
      const ev = e.triangularValue[k] as number;
      if (!closeEnough(av, ev)) {
        divs.push({
          field: `${prefix}.triangularValue.${k}`,
          expected: ev,
          actual: av,
          diff: Math.abs(av - ev),
        });
      }
    }
    for (const k of ['positive', 'average', 'negative'] as const) {
      const av = a.normalizedValue[k];
      const ev = e.normalizedValue[k] as number;
      if (!closeEnough(av, ev)) {
        divs.push({
          field: `${prefix}.normalizedValue.${k}`,
          expected: ev,
          actual: av,
          diff: Math.abs(av - ev),
        });
      }
    }
    if (!closeEnough(a.score, e.score as number)) {
      divs.push({
        field: `${prefix}.score`,
        expected: e.score,
        actual: a.score,
        diff: Math.abs(a.score - (e.score as number)),
      });
    }
    if (a.rank !== (e.rank as number)) {
      divs.push({ field: `${prefix}.rank`, expected: e.rank, actual: a.rank });
    }
  }
  return divs;
}

function compareCriticalConstraints(
  actual: CriticalConstraintM[],
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  expected: any[],
  label: string,
): Divergence[] {
  const divs: Divergence[] = [];

  if (actual.length !== expected.length) {
    divs.push({ field: `${label}.length`, expected: expected.length, actual: actual.length });
    return divs;
  }

  for (let i = 0; i < actual.length; i++) {
    const a = actual[i];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const e = expected[i] as any;
    const prefix = `${label}[${i}]`;

    if (a.constraintIndex !== (e.constraintIndex as number)) {
      divs.push({
        field: `${prefix}.constraintIndex`,
        expected: e.constraintIndex,
        actual: a.constraintIndex,
      });
    }
    if (a.kind !== (e.kind as string)) {
      divs.push({ field: `${prefix}.kind`, expected: e.kind, actual: a.kind });
    }
    if (a.eliminated !== (e.eliminated as number)) {
      divs.push({ field: `${prefix}.eliminated`, expected: e.eliminated, actual: a.eliminated });
    }
    if (a.total !== (e.total as number)) {
      divs.push({ field: `${prefix}.total`, expected: e.total, actual: a.total });
    }
    if (a.redundant !== (e.redundant as boolean)) {
      divs.push({ field: `${prefix}.redundant`, expected: e.redundant, actual: a.redundant });
    }
  }
  return divs;
}

// ---------------------------------------------------------------------------
// Main runner
// ---------------------------------------------------------------------------
export function runConformance(): { passed: number; failed: number; divergences: Divergence[][] } {
  const scenarioFiles = fs
    .readdirSync(SCENARIOS_DIR)
    .filter((f) => f.endsWith('.json'))
    .sort();

  let passed = 0;
  let failed = 0;
  const allDivergences: Divergence[][] = [];

  for (const file of scenarioFiles) {
    const name = path.basename(file, '.json'); // e.g. "sas"
    const scenarioPath = path.join(SCENARIOS_DIR, file);

    console.log(`\n--- ${name} ---`);

    let scenario;
    try {
      scenario = loadScenario(scenarioPath);
    } catch (e) {
      console.error(`  FAIL: loadScenario threw: ${e}`);
      failed++;
      allDivergences.push([{ field: 'loadScenario', expected: 'success', actual: String(e) }]);
      continue;
    }

    const candidates = solve(scenario);
    const critDecs = criticalDecisions(scenario, candidates);
    const critConsts = criticalConstraints(scenario);

    const divs: Divergence[] = [];

    // Compare candidates
    const expectedCandidatesPath = path.join(EXPECTED_DIR, `${name}.candidates.json`);
    if (fs.existsSync(expectedCandidatesPath)) {
      const expectedCandidates = JSON.parse(fs.readFileSync(expectedCandidatesPath, 'utf-8'));
      divs.push(...compareCandidates(candidates, expectedCandidates, `${name}.candidates`));
    }

    // Compare critical decisions
    const expectedCritDecsPath = path.join(EXPECTED_DIR, `${name}.critical-decisions.json`);
    if (fs.existsSync(expectedCritDecsPath)) {
      const expectedCritDecs = JSON.parse(fs.readFileSync(expectedCritDecsPath, 'utf-8'));
      divs.push(
        ...compareCriticalDecisions(critDecs, expectedCritDecs, `${name}.critical-decisions`),
      );
    }

    // Compare critical constraints
    const expectedCritConstsPath = path.join(EXPECTED_DIR, `${name}.critical-constraints.json`);
    if (fs.existsSync(expectedCritConstsPath)) {
      const expectedCritConsts = JSON.parse(fs.readFileSync(expectedCritConstsPath, 'utf-8'));
      divs.push(
        ...compareCriticalConstraints(
          critConsts,
          expectedCritConsts,
          `${name}.critical-constraints`,
        ),
      );
    }

    if (divs.length === 0) {
      console.log(`  PASS`);
      passed++;
    } else {
      console.error(`  FAIL: ${divs.length} divergence(s)`);
      for (const d of divs) {
        if (d.diff !== undefined) {
          console.error(
            `    ${d.field}: expected=${d.expected}, actual=${d.actual}, diff=${d.diff}`,
          );
        } else {
          console.error(
            `    ${d.field}: expected=${JSON.stringify(d.expected)}, actual=${JSON.stringify(d.actual)}`,
          );
        }
      }
      failed++;
      allDivergences.push(divs);
    }
  }

  console.log(`\nResults: ${passed} passed, ${failed} failed`);
  return { passed, failed, divergences: allDivergences };
}

// ---------------------------------------------------------------------------
// CLI entry point — only runs when this file is executed directly
// ---------------------------------------------------------------------------
// Detect if this file is the entry point (not imported as a module)
const isMain =
  typeof process !== 'undefined' &&
  // In ESM, import.meta.url is the canonical way; for tsx/node direct execution
  // process.argv[1] will match this file's path
  process.argv[1] !== undefined &&
  (process.argv[1].endsWith('runner.ts') ||
    process.argv[1].endsWith('runner.js') ||
    process.argv[1].includes('conformance/runner'));

if (isMain) {
  const { failed: exitCode } = runConformance();
  process.exit(exitCode > 0 ? 1 : 0);
}

File: langs/python/src/guidearch/conformance/runner.py
"""Conformance runner — compares solver output against spec/conformance/expected/*.

CLI entry point: python -m guidearch.conformance.runner
  Exits 0 on conformance, 1 on divergence.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from guidearch.models.scenario_loader import load_scenario
from guidearch.models.topsis.critical_constraints import critical_constraints
from guidearch.models.topsis.critical_decisions import critical_decisions
from guidearch.models.topsis.solve import solve
from guidearch.output.serialize import (
    candidates_to_dict,
    critical_constraints_to_dict,
    critical_decisions_to_dict,
)

# Path to the conformance corpus. The dev / in-tree walk works the same way
# scenario_loader._load_default_schema does (parents[5] is the repo root from
# this module). Wheel-installed callers won't have spec/ alongside the package;
# scenarios + expected files are not bundled into the wheel, so wheel users
# must point GUIDEARCH_SPEC_DIR at a checkout. The env-var override keeps the
# CLI usable from outside the repo without requiring an in-tree check-in.
_DEFAULT_SPEC_DIR = Path(__file__).parents[5] / "spec" / "conformance"
_SPEC_DIR = Path(__import__("os").environ.get("GUIDEARCH_SPEC_DIR", str(_DEFAULT_SPEC_DIR)))


def _load_abs_tol(spec_dir: Path) -> float:
    """Read the scalar absolute tolerance from spec/conformance/tolerances.json.

    All scalar fields in the corpus share the same absolute tolerance at
    v1.0; tolerances.json was the single source of truth for TS already.
    Falling back to 1e-9 keeps the runner usable when the corpus is missing
    (e.g. wheel install without GUIDEARCH_SPEC_DIR).
    """
    tol_path = spec_dir / "tolerances.json"
    if not tol_path.exists():
        return 1e-9
    data = json.loads(tol_path.read_text(encoding="utf-8"))
    return float(data["candidates"]["score"]["absolute"])


_ABS_TOL = _load_abs_tol(_SPEC_DIR)


@dataclass
class Difference:
    """A single conformance failure."""

    scenario: str
    kind: str  # 'candidates' | 'critical-decisions' | 'critical-constraints'
    index: int
    field: str
    expected: object
    actual: object

    def __str__(self) -> str:
        return (
            f"[{self.scenario}] {self.kind}[{self.index}].{self.field}: "
            f"expected={self.expected!r}, actual={self.actual!r}"
        )


def _float_close(a: float, b: float, tol: float = _ABS_TOL) -> bool:
    return abs(a - b) <= tol


def _compare_tfm(
    exp: dict[str, Any],
    act: dict[str, Any],
    diffs: list[Difference],
    scenario: str,
    kind: str,
    idx: int,
    prefix: str,
) -> None:
    for key in ("lower", "modal", "upper"):
        if not _float_close(exp[key], act[key]):
            diffs.append(Difference(scenario, kind, idx, f"{prefix}.{key}", exp[key], act[key]))


def _compare_nfm(
    exp: dict[str, Any],
    act: dict[str, Any],
    diffs: list[Difference],
    scenario: str,
    kind: str,
    idx: int,
    prefix: str,
) -> None:
    for key in ("positive", "average", "negative"):
        if not _float_close(exp[key], act[key]):
            diffs.append(Difference(scenario, kind, idx, f"{prefix}.{key}", exp[key], act[key]))


def _compare_candidates(
    exp_list: list[dict[str, Any]],
    act_list: list[dict[str, Any]],
    diffs: list[Difference],
    scenario: str,
) -> None:
    kind = "candidates"
    if len(exp_list) != len(act_list):
        diffs.append(Difference(scenario, kind, -1, "count", len(exp_list), len(act_list)))
        return
    for i, (exp, act) in enumerate(zip(exp_list, act_list, strict=True)):
        if exp["alternativeIds"] != act["alternativeIds"]:
            diffs.append(
                Difference(
                    scenario,
                    kind,
                    i,
                    "alternativeIds",
                    exp["alternativeIds"],
                    act["alternativeIds"],
                )
            )
        _compare_tfm(
            exp["triangularValue"],
            act["triangularValue"],
            diffs,
            scenario,
            kind,
            i,
            "triangularValue",
        )
        _compare_nfm(
            exp["normalizedValue"],
            act["normalizedValue"],
            diffs,
            scenario,
            kind,
            i,
            "normalizedValue",
        )
        if not _float_close(exp["score"], act["score"]):
            diffs.append(Difference(scenario, kind, i, "score", exp["score"], act["score"]))
        if exp["rank"] != act["rank"]:
            diffs.append(Difference(scenario, kind, i, "rank", exp["rank"], act["rank"]))


def _compare_critical_decisions(
    exp_list: list[dict[str, Any]],
    act_list: list[dict[str, Any]],
    diffs: list[Difference],
    scenario: str,
) -> None:
    kind = "critical-decisions"
    if len(exp_list) != len(act_list):
        diffs.append(Difference(scenario, kind, -1, "count", len(exp_list), len(act_list)))
        return
    for i, (exp, act) in enumerate(zip(exp_list, act_list, strict=True)):
        if exp["decisionId"] != act["decisionId"]:
            diffs.append(
                Difference(
                    scenario,
                    kind,
                    i,
                    "decisionId",
                    exp["decisionId"],
                    act["decisionId"],
                )
            )
        _compare_tfm(
            exp["triangularValue"],
            act["triangularValue"],
            diffs,
            scenario,
            kind,
            i,
            "triangularValue",
        )
        _compare_nfm(
            exp["normalizedValue"],
            act["normalizedValue"],
            diffs,
            scenario,
            kind,
            i,
            "normalizedValue",
        )
        if not _float_close(exp["score"], act["score"]):
            diffs.append(Difference(scenario, kind, i, "score", exp["score"], act["score"]))
        if exp["rank"] != act["rank"]:
            diffs.append(Difference(scenario, kind, i, "rank", exp["rank"], act["rank"]))


def _compare_critical_constraints(
    exp_list: list[dict[str, Any]],
    act_list: list[dict[str, Any]],
    diffs: list[Difference],
    scenario: str,
) -> None:
    kind = "critical-constraints"
    if len(exp_list) != len(act_list):
        diffs.append(Difference(scenario, kind, -1, "count", len(exp_list), len(act_list)))
        return
    for i, (exp, act) in enumerate(zip(exp_list, act_list, strict=True)):
        for field in ("constraintIndex", "kind", "eliminated", "total", "redundant"):
            if exp[field] != act[field]:
                diffs.append(Difference(scenario, kind, i, field, exp[field], act[field]))


def run_conformance(spec_dir: Path = _SPEC_DIR) -> list[Difference]:
    """Run conformance for all scenarios in spec_dir/scenarios/*.json.

    Compares against spec_dir/expected/<name>.{candidates,critical-decisions,
    critical-constraints}.json, honoring tolerances.json.
    """
    scenarios_dir = spec_dir / "scenarios"
    expected_dir = spec_dir / "expected"

    all_diffs: list[Difference] = []

    for scenario_path in sorted(scenarios_dir.glob("*.json")):
        name = scenario_path.stem.lower()

        scenario = load_scenario(scenario_path)
        candidates = solve(scenario)
        cd = critical_decisions(scenario, candidates)
        cc = critical_constraints(scenario)

        actual_candidates = candidates_to_dict(candidates)
        actual_cd = critical_decisions_to_dict(cd)
        actual_cc = critical_constraints_to_dict(cc)

        cand_exp_path = expected_dir / f"{name}.candidates.json"
        cd_exp_path = expected_dir / f"{name}.critical-decisions.json"
        cc_exp_path = expected_dir / f"{name}.critical-constraints.json"

        if cand_exp_path.exists():
            exp_candidates = json.loads(cand_exp_path.read_text(encoding="utf-8"))
            _compare_candidates(exp_candidates, actual_candidates, all_diffs, name)
        else:
            all_diffs.append(
                Difference(
                    name,
                    "candidates",
                    -1,
                    "expected_file",
                    str(cand_exp_path),
                    "missing",
                )
            )

        if cd_exp_path.exists():
            exp_cd = json.loads(cd_exp_path.read_text(encoding="utf-8"))
            _compare_critical_decisions(exp_cd, actual_cd, all_diffs, name)
        else:
            all_diffs.append(
                Difference(
                    name,
                    "critical-decisions",
                    -1,
                    "expected_file",
                    str(cd_exp_path),
                    "missing",
                )
            )

        if cc_exp_path.exists():
            exp_cc = json.loads(cc_exp_path.read_text(encoding="utf-8"))
            _compare_critical_constraints(exp_cc, actual_cc, all_diffs, name)
        else:
            all_diffs.append(
                Difference(
                    name,
                    "critical-constraints",
                    -1,
                    "expected_file",
                    str(cc_exp_path),
                    "missing",
                )
            )

    return all_diffs


def main() -> int:
    diffs = run_conformance()
    if not diffs:
        print("PASS: all conformance checks passed")
        return 0
    print(f"FAIL: {len(diffs)} conformance difference(s):")
    for d in diffs:
        print(f"  {d}")
    return 1


if __name__ == "__main__":
    sys.exit(main())
